### ISSUE TITLE
Download media only when the client need them

### ISSUE THREAD

**Issue by Emojigit** [minetest/minetest#12713]
## Problem

The process of downloading media is VERY SLOW, esp. for proxy users. That's because Minetest client pull all the media used from the server, regardless they are needed or not. 

## Solutions

A solution is only to download necessary media when the client require them to render out something. Initially, the client only download the texture of hand and the crack textures.

There is an unsolved problem though: client have to wait a lot of time before rendering to finish then.

**Comment by NathanSalapat** [comment]
The media files from a server are usually stored locally on the client computer in a cache someplace. It's normally only necessary to download all the media the first time joining. Any future joins should only download any new/modified media.

**Comment by Emojigit** [comment]
Yeah, I know, though the ONE FIRST time is also slow enough to annoy a player...

**Comment by appgurueu** [comment]
You can (mostly) implement this yourself using dynamic media. You just have to be careful to make sure that nothing using yet-unsent textures is sent to the player before they have received the relevant textures. You will need to use a proxy for that.

**Comment by SmallJoker** [comment]
It's easier said than done. Situation analysis and implementation ideas:

1. Node definition and media (currently) are received on client join
2. `ContentFeatures::updateTextures` -> `fillTileAttribs` -> `tsrc->getTextureForMesh` get an unique texture ID based on the node definition tiles. When a texture is not present, a dummy one will be generated and used. Hence you'd have to catch unknown textures in `TextureSource::getTextureId` -> `TextureSource::generateTexture`, assign a special texture ID to `TileLayer::texture_id`.
3. In-game, only `MapBlockMesh::MapBlockMesh` could notice the missing texture. Hence trigger a request there.
4. ^ within, detect the special texture ID and request the texture from the server, which then would have to trigger an update for all affected mapblock meshes once it arrives.

.. good luck with that.

**Comment by Desour** [comment]
For sounds at least, much (or most) of the loading time doesn't come from the download, but from the media loading (i.e. ogg vorbis decompression). (I'm working on a PR to improve the situation, but it's not yet completed.)

16x16 textures are most likely no bottleneck, I'd guess. Just think of all the 16x16x16 node mapblocks that have to be sent all the time.

**Comment by rubenwardy** [comment]
Are you using Android or an arm-based device? There's a known issue that effects loading

The problem is that most media needs to be available immediately when requested. So sending all the media during load prevents jitter and issues whilst playing.

Minetest's dynamic media API allows mods to send media just in time.

It would be nice to support music streaming too. Unlike sound effects, it doesn't matter if music is delayed for a few seconds and it's more predictable so can be preloaded before the previous track ends.

TL;DR: this issue should be closed, any slowness in

**Comment by Emojigit** [comment]
> Are you using Android or an arm-based device? There's a known issue that effects loading

In my case, the player is connecting though a proxy. I think it would be better to discretely download the media.

> 16x16 textures are most likely no bottleneck, I'd guess.

Yes, though a lot of mods are using 32x32 (e.g. redwood wood of ethereal), and there are A LOOOOOOOT of meshes (advtrains, moretrains, etc.) in my case.
